---
title: "RNA/Protein Overlap"
author: "Simon Haugaard"
date: "2024-06-24"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Library
```{r}
library(ggvenn)
if (!require("pacman")) install.packages("pacman")
pacman::p_load("edgeR")
pacman::p_load("readr")
pacman::p_load("readxl")
pacman::p_load("biomaRt")
pacman::p_load("magrittr")
pacman::p_load("tibble")
pacman::p_load("stringr")
pacman::p_load("ggplot2")
pacman::p_load("data.table")
pacman::p_load("ggplot2", "patchwork")
pacman::p_load("openxlsx")
library(dplyr)
library(missForest)
library(RColorBrewer)
library(limma)
library(DEqMS)
library(preprocessCore)
library(DEP)
library(ggvenn)

```


#Read files
```{r}
#Metformin_vs_placebo
RNA <- fread("C:/Users/nks882/OneDrive - University of Copenhagen/Documents/GitHub/Horse_AF_project/RNA-seq/analysis/01_dge/output/dge_results.tsv.gz")
Protein <- fread("C:/Users/nks882/OneDrive - University of Copenhagen/Documents/GitHub/Horse_AF_project/Proteomics/Terminal/analysis/01_dge/output/dge_results.tsv.gz")
```

#Seperate into contrasts
```{r}
# Separate into contrasts and select the appropriate columns
# RNA filtering based on adj. P values (or raw p.values for LA)
metformin_vs_placebo_RNA_sig_up_LA <- RNA[Contrast == "met_vs_placebo_LA" & logFC > 0 & P.Value < 0.05, .(GENENAME)]
metformin_vs_placebo_RNA_sig_up_RA <- RNA[Contrast == "met_vs_placebo_RA" & logFC > 0 & adj.P.Val < 0.05, .(GENENAME)]
metformin_vs_placebo_RNA_sig_down_LA <- RNA[Contrast == "met_vs_placebo_LA" & logFC < 0 & P.Value < 0.05, .(GENENAME)]
metformin_vs_placebo_RNA_sig_down_RA <- RNA[Contrast == "met_vs_placebo_RA" & logFC < 0 & adj.P.Val < 0.05, .(GENENAME)]

# Protein filtering with ranking and top 100 selection based on p.value first and logFC second
metformin_vs_placebo_protein_sig_up_LA <- Protein[Contrast == "Metformin_vs_AF_LA" & logFC > 0, .(GeneName, P.Value, logFC)]
metformin_vs_placebo_protein_sig_up_LA <- metformin_vs_placebo_protein_sig_up_LA[order(P.Value, -logFC)][1:100, .(GeneName)]

metformin_vs_placebo_protein_sig_up_RA <- Protein[Contrast == "Metformin_vs_AF_RA" & logFC > 0, .(GeneName, P.Value, logFC)]
metformin_vs_placebo_protein_sig_up_RA <- metformin_vs_placebo_protein_sig_up_RA[order(P.Value, -logFC)][1:100, .(GeneName)]

metformin_vs_placebo_protein_sig_down_LA <- Protein[Contrast == "Metformin_vs_AF_LA" & logFC < 0, .(GeneName, P.Value, logFC)]
metformin_vs_placebo_protein_sig_down_LA <- metformin_vs_placebo_protein_sig_down_LA[order(P.Value, logFC)][1:100, .(GeneName)]

metformin_vs_placebo_protein_sig_down_RA <- Protein[Contrast == "Metformin_vs_AF_RA" & logFC < 0, .(GeneName, P.Value, logFC)]
metformin_vs_placebo_protein_sig_down_RA <- metformin_vs_placebo_protein_sig_down_RA[order(P.Value, logFC)][1:100, .(GeneName)]

placebo_vs_sham_RNA_sig_up_LA <- RNA[Contrast == "AF_vs_sham_LA" & logFC > 0 & adj.P.Val < 0.05, .(GENENAME)]
placebo_vs_sham_RNA_sig_up_RA <- RNA[Contrast == "AF_vs_sham_RA" & logFC > 0 & adj.P.Val < 0.05, .(GENENAME)]
placebo_vs_sham_RNA_sig_down_LA <- RNA[Contrast == "AF_vs_sham_LA" & logFC < 0 & adj.P.Val < 0.05, .(GENENAME)]
placebo_vs_sham_RNA_sig_down_RA <- RNA[Contrast == "AF_vs_sham_RA" & logFC < 0 & adj.P.Val < 0.05, .(GENENAME)]

# Protein filtering with ranking and top 100 selection
placebo_vs_sham_protein_sig_up_LA <- Protein[Contrast == "AF_vs_Sham_LA" & logFC > 0, .(GeneName, P.Value, logFC)]
placebo_vs_sham_protein_sig_up_LA <- placebo_vs_sham_protein_sig_up_LA[order(P.Value, -logFC)][1:100, .(GeneName)]

placebo_vs_sham_protein_sig_up_RA <- Protein[Contrast == "AF_vs_Sham_RA" & logFC > 0, .(GeneName, P.Value, logFC)]
placebo_vs_sham_protein_sig_up_RA <- placebo_vs_sham_protein_sig_up_RA[order(P.Value, -logFC)][1:100, .(GeneName)]

placebo_vs_sham_protein_sig_down_LA <- Protein[Contrast == "AF_vs_Sham_LA" & logFC < 0, .(GeneName, P.Value, logFC)]
placebo_vs_sham_protein_sig_down_LA <- placebo_vs_sham_protein_sig_down_LA[order(P.Value, logFC)][1:100, .(GeneName)]

placebo_vs_sham_protein_sig_down_RA <- Protein[Contrast == "AF_vs_Sham_RA" & logFC < 0, .(GeneName, P.Value, logFC)]
placebo_vs_sham_protein_sig_down_RA <- placebo_vs_sham_protein_sig_down_RA[order(P.Value, logFC)][1:100, .(GeneName)]

# Disease effect
disease_effect_protein_down <- Protein[Contrast == "AverageDiseaseEffect" & logFC < 0, .(GeneName, P.Value, logFC)]
disease_effect_protein_down <- disease_effect_protein_down[order(P.Value, logFC)][1:100, .(GeneName)]

disease_effect_protein_up <- Protein[Contrast == "AverageDiseaseEffect" & logFC > 0, .(GeneName, P.Value, logFC)]
disease_effect_protein_up <- disease_effect_protein_up[order(P.Value, -logFC)][1:100, .(GeneName)]

disease_effect_RNA_down <- RNA[Contrast == "AverageDiseaseEffect" & logFC < 0 & adj.P.Val < 0.05, .(GENENAME)]
disease_effect_RNA_up <- RNA[Contrast == "AverageDiseaseEffect" & logFC > 0 & adj.P.Val < 0.05, .(GENENAME)]

# Treatment effect
treatment_effect_protein_down <- Protein[Contrast == "AverageTreatmentEffect" & logFC < 0, .(GeneName, P.Value, logFC)]
treatment_effect_protein_down <- treatment_effect_protein_down[order(P.Value, logFC)][1:100, .(GeneName)]

treatment_effect_protein_up <- Protein[Contrast == "AverageTreatmentEffect" & logFC > 0, .(GeneName, P.Value, logFC)]
treatment_effect_protein_up <- treatment_effect_protein_up[order(P.Value, -logFC)][1:100, .(GeneName)]

treatment_effect_RNA_down <- RNA[Contrast == "AverageTreatmentEffect" & logFC < 0 & adj.P.Val < 0.05, .(GENENAME)]
treatment_effect_RNA_up <- RNA[Contrast == "AverageTreatmentEffect" & logFC > 0 & adj.P.Val < 0.05, .(GENENAME)]

```

#Metformin-RA
```{r}
# Ensure they are vectors
metformin_vs_placebo_RNA_sig_up_RA <- metformin_vs_placebo_RNA_sig_up_RA$GENENAME
metformin_vs_placebo_RNA_sig_down_RA <- metformin_vs_placebo_RNA_sig_down_RA$GENENAME
metformin_vs_placebo_protein_sig_up_RA <- metformin_vs_placebo_protein_sig_up_RA$GeneName
metformin_vs_placebo_protein_sig_down_RA <- metformin_vs_placebo_protein_sig_down_RA$GeneName

# Create a list of data sets
gene_lists <- list(
  RNA_Up = metformin_vs_placebo_RNA_sig_up_RA,
  RNA_Down = metformin_vs_placebo_RNA_sig_down_RA,
  Protein_Up = metformin_vs_placebo_protein_sig_up_RA,
  Protein_Down = metformin_vs_placebo_protein_sig_down_RA
)

# Generate Venn diagram
ggvenn(gene_lists, fill_color = c("#E69F00", "#56B4E9", "#009E73", "#F0E442"))

# Compute overlaps
overlap_sig_up <- intersect(metformin_vs_placebo_RNA_sig_up_RA, metformin_vs_placebo_protein_sig_up_RA)
overlap_sig_down <- intersect(metformin_vs_placebo_RNA_sig_down_RA, metformin_vs_placebo_protein_sig_down_RA)
non_overlap1 <- intersect(metformin_vs_placebo_RNA_sig_down_RA, metformin_vs_placebo_protein_sig_up_RA)
non_overlap2 <- intersect(metformin_vs_placebo_RNA_sig_up_RA, metformin_vs_placebo_protein_sig_down_RA)

# Create a list with the overlaps
overlap_list <- list(
  Overlap_Sig_Up = overlap_sig_up,
  Overlap_Sig_Down = overlap_sig_down, 
  Down_rna_up_protein = non_overlap1,
  Up_rna_down_protein = non_overlap2
)

# Print the overlap list in a comma-separated format
cat("Overlap_Sig_Up:", paste(overlap_list$Overlap_Sig_Up, collapse = ", "), "\n")
cat("Overlap_Sig_Down:", paste(overlap_list$Overlap_Sig_Down, collapse = ", "), "\n")
cat("Down_rna_up_protein:", paste(overlap_list$Down_rna_up_protein, collapse = ", "), "\n")
    # Relates to fatty acids, acetylation, mitochondria, oxidative stress
cat("Up_rna_down_protein:", paste(overlap_list$Up_rna_down_protein, collapse = ", "), "\n")
    # Ubiquitin, prtoeosome, lysosome, 

```
#Metformin LA
```{r}
# Ensure they are vectors
metformin_vs_placebo_RNA_sig_up_LA <- metformin_vs_placebo_RNA_sig_up_LA$GENENAME
metformin_vs_placebo_RNA_sig_down_LA <- metformin_vs_placebo_RNA_sig_down_LA$GENENAME
metformin_vs_placebo_protein_sig_up_LA <- metformin_vs_placebo_protein_sig_up_LA$GeneName
metformin_vs_placebo_protein_sig_down_LA <- metformin_vs_placebo_protein_sig_down_LA$GeneName

# Create a list of data sets
gene_lists <- list(
  RNA_Up = metformin_vs_placebo_RNA_sig_up_LA,
  RNA_Down = metformin_vs_placebo_RNA_sig_down_LA,
  Protein_Up = metformin_vs_placebo_protein_sig_up_LA,
  Protein_Down = metformin_vs_placebo_protein_sig_down_LA
)

# Generate Venn diagram
ggvenn(gene_lists, fill_color = c("#E69F00", "#56B4E9", "#009E73", "#F0E442"))

# Compute overlaps
overlap_sig_up <- intersect(metformin_vs_placebo_RNA_sig_up_LA, metformin_vs_placebo_protein_sig_up_LA)
overlap_sig_down <- intersect(metformin_vs_placebo_RNA_sig_down_LA, metformin_vs_placebo_protein_sig_down_LA)

# Create a list with the overlaps
overlap_list <- list(
  Overlap_Sig_Up = overlap_sig_up,
  Overlap_Sig_Down = overlap_sig_down
)

# Print the overlap list in a comma-separated format
cat("Overlap_Sig_Up:", paste(overlap_list$Overlap_Sig_Up, collapse = ", "), "\n")
cat("Overlap_Sig_Down:", paste(overlap_list$Overlap_Sig_Down, collapse = ", "), "\n")

```

#AF-RA
```{r}
# Ensure they are vectors
placebo_vs_sham_RNA_sig_up_RA <- placebo_vs_sham_RNA_sig_up_RA$GENENAME
placebo_vs_sham_RNA_sig_down_RA <- placebo_vs_sham_RNA_sig_down_RA$GENENAME
placebo_vs_sham_protein_sig_up_RA <- placebo_vs_sham_protein_sig_up_RA$GeneName
placebo_vs_sham_protein_sig_down_RA <- placebo_vs_sham_protein_sig_down_RA$GeneName

# Create a list of data sets
gene_lists <- list(
  RNA_Up = placebo_vs_sham_RNA_sig_up_RA,
  RNA_Down = placebo_vs_sham_RNA_sig_down_RA,
  Protein_Up = placebo_vs_sham_protein_sig_up_RA,
  Protein_Down = placebo_vs_sham_protein_sig_down_RA
)

# Generate Venn diagram
ggvenn(gene_lists, fill_color = c("#E69F00", "#56B4E9", "#009E73", "#F0E442"))

# Compute overlaps
overlap_sig_up <- intersect(placebo_vs_sham_RNA_sig_up_RA, placebo_vs_sham_protein_sig_up_RA)
overlap_sig_down <- intersect(placebo_vs_sham_RNA_sig_down_RA, placebo_vs_sham_protein_sig_down_RA)

# Create a list with the overlaps
overlap_list <- list(
  Overlap_Sig_Up = overlap_sig_up,
  Overlap_Sig_Down = overlap_sig_down
)

# Print the overlap list in a comma-separated format
cat("Overlap_Sig_Up:", paste(overlap_list$Overlap_Sig_Up, collapse = ", "), "\n")
cat("Overlap_Sig_Down:", paste(overlap_list$Overlap_Sig_Down, collapse = ", "), "\n")

```
#AF-LA 
```{r}
# Ensure they are vectors
placebo_vs_sham_RNA_sig_up_LA <- placebo_vs_sham_RNA_sig_up_LA$GENENAME
placebo_vs_sham_RNA_sig_down_LA <- placebo_vs_sham_RNA_sig_down_LA$GENENAME
placebo_vs_sham_protein_sig_up_LA <- placebo_vs_sham_protein_sig_up_LA$GeneName
placebo_vs_sham_protein_sig_down_LA <- placebo_vs_sham_protein_sig_down_LA$GeneName

# Create a list of data sets
gene_lists <- list(
  RNA_up= placebo_vs_sham_RNA_sig_up_LA,
  RNA_down= placebo_vs_sham_RNA_sig_down_LA,
  Protein_up = placebo_vs_sham_protein_sig_up_LA,
  Protein_down = placebo_vs_sham_protein_sig_down_LA
)

# Generate Venn diagram
ggvenn(gene_lists, fill_color = c("#E69F00", "#56B4E9", "#009E73", "#F0E442"))

# Compute overlaps
overlap_sig_up <- intersect(placebo_vs_sham_RNA_sig_up_LA, placebo_vs_sham_protein_sig_up_LA)
overlap_sig_down <- intersect(placebo_vs_sham_RNA_sig_down_LA, placebo_vs_sham_protein_sig_down_LA)

# Create a list with the overlaps
overlap_list <- list(
  Overlap_Sig_Up = overlap_sig_up,
  Overlap_Sig_Down = overlap_sig_down
)

# Print the overlap list in a comma-separated format
cat("Overlap_Sig_Up:", paste(overlap_list$Overlap_Sig_Up, collapse = ", "), "\n")
cat("Overlap_Sig_Down:", paste(overlap_list$Overlap_Sig_Down, collapse = ", "), "\n")

```

#Disease
```{r}
# Ensure they are vectors
disease_effect_RNA_up <- disease_effect_RNA_up$GENENAME
disease_effect_RNA_down <- disease_effect_RNA_down$GENENAME
disease_effect_protein_up <- disease_effect_protein_up$GeneName
disease_effect_protein_down <- disease_effect_protein_down$GeneName

# Create a list of data sets
gene_lists <- list(
  RNA_up= disease_effect_RNA_up,
  RNA_down= disease_effect_RNA_down,
  Protein_up = disease_effect_protein_up,
  Protein_down = disease_effect_protein_down
)

# Generate Venn diagram
ggvenn(gene_lists, fill_color = c("#E69F00", "#56B4E9", "#009E73", "#F0E442"))

# Compute overlaps
overlap_sig_up <- intersect(disease_effect_RNA_up, disease_effect_protein_up)
overlap_sig_down <- intersect(disease_effect_RNA_down, disease_effect_protein_down)

# Create a list with the overlaps
overlap_list <- list(
  Overlap_Sig_Up = overlap_sig_up,
  Overlap_Sig_Down = overlap_sig_down
)

# Print the overlap list in a comma-separated format
cat("Overlap_Sig_Up:", paste(overlap_list$Overlap_Sig_Up, collapse = ", "), "\n")
cat("Overlap_Sig_Down:", paste(overlap_list$Overlap_Sig_Down, collapse = ", "), "\n")
```


#Treatment
```{r}
# Ensure they are vectors
treatment_effect_RNA_up <- treatment_effect_RNA_up$GENENAME
treatment_effect_RNA_down <- treatment_effect_RNA_down$GENENAME
treatment_effect_protein_up <- treatment_effect_protein_up$GeneName
treatment_effect_protein_down <- treatment_effect_protein_down$GeneName

# Create a list of data sets
gene_lists <- list(
  RNA_up= treatment_effect_RNA_up,
  RNA_down= treatment_effect_RNA_down,
  Protein_up = treatment_effect_protein_up,
  Protein_down = treatment_effect_protein_down
)

# Generate Venn diagram
ggvenn(gene_lists, fill_color = c("#E69F00", "#56B4E9", "#009E73", "#F0E442"))

# Compute overlaps
overlap_sig_up <- intersect(treatment_effect_RNA_up, treatment_effect_protein_up)
overlap_sig_down <- intersect(treatment_effect_RNA_down, treatment_effect_protein_down)

# Create a list with the overlaps
overlap_list <- list(
  Overlap_Sig_Up = overlap_sig_up,
  Overlap_Sig_Down = overlap_sig_down
)

# Print the overlap list in a comma-separated format
cat("Overlap_Sig_Up:", paste(overlap_list$Overlap_Sig_Up, collapse = ", "), "\n")
cat("Overlap_Sig_Down:", paste(overlap_list$Overlap_Sig_Down, collapse = ", "), "\n")
```


#Supplementary Figure - Biatrial Effect of AF - what is upregulated?
```{r}
placebo_vs_sham_protein_sig_up_LA <- Protein[Contrast == "AF_vs_Sham_LA" & logFC > 0 & P.Value < 0.05, .(GeneName)]$GeneName
placebo_vs_sham_protein_sig_down_LA <- Protein[Contrast == "AF_vs_Sham_LA" & logFC < 0 & P.Value < 0.05, .(GeneName)]$GeneName

placebo_vs_sham_protein_sig_up_RA <- Protein[Contrast == "AF_vs_Sham_RA" & logFC > 0 & P.Value < 0.05, .(GeneName)]$GeneName
placebo_vs_sham_protein_sig_down_RA <- Protein[Contrast == "AF_vs_Sham_RA" & logFC < 0 & P.Value < 0.05, .(GeneName)]$GeneName



gene_lists <- list(
  "RNA_Up_RA" = placebo_vs_sham_RNA_sig_up_RA,
  "Protein_Up_RA" = placebo_vs_sham_protein_sig_up_RA,
  "RNA_Up_LA" = placebo_vs_sham_RNA_sig_up_LA,
  "Protein_Up_LA" = placebo_vs_sham_protein_sig_up_LA)

# Generate the Venn diagram with custom colors
ggvenn(gene_lists, fill_color = c("#E69F00", "#56B4E9", "#009E73", "#F0E442"))

# Compute the overlap between all four circles
overlap_all <- Reduce(intersect, list(
  placebo_vs_sham_RNA_sig_up_RA,
  placebo_vs_sham_protein_sig_up_RA,
  placebo_vs_sham_RNA_sig_up_LA,
  placebo_vs_sham_protein_sig_up_LA
))

# Print the names of the genes in the overlap
cat("Overlap between all circles (RA RNA Up, RA Protein Up, LA RNA Up, LA Protein Up):", 
    paste(overlap_all, collapse = ", "), "\n")


```

#Biatrial Effect of AF - what is downregulated?
```{r}
# For Downregulated Genes/Proteins
placebo_vs_sham_protein_sig_down_LA <- Protein[Contrast == "AF_vs_Sham_LA" & logFC < 0 & P.Value < 0.05, .(GeneName)]$GeneName
placebo_vs_sham_protein_sig_down_RA <- Protein[Contrast == "AF_vs_Sham_RA" & logFC < 0 & P.Value < 0.05, .(GeneName)]$GeneName

gene_lists_down <- list(
  "RNA_Down_RA" = placebo_vs_sham_RNA_sig_down_RA,
  "Protein_Down_RA" = placebo_vs_sham_protein_sig_down_RA,
  "RNA_Down_LA" = placebo_vs_sham_RNA_sig_down_LA,
  "Protein_Down_LA" = placebo_vs_sham_protein_sig_down_LA
)

# Generate the Venn diagram for downregulated overlaps with custom colors
ggvenn(gene_lists_down, fill_color = c("#E69F00", "#56B4E9", "#009E73", "#F0E442"))

# Compute the overlap between all four circles for downregulated genes/proteins
overlap_all_down <- Reduce(intersect, list(
  placebo_vs_sham_RNA_sig_down_RA,
  placebo_vs_sham_protein_sig_down_RA,
  placebo_vs_sham_RNA_sig_down_LA,
  placebo_vs_sham_protein_sig_down_LA
))

# Print the names of the genes in the overlap for downregulated genes/proteins
cat("Overlap between all circles (Downregulated in RA RNA, RA Protein, LA RNA, LA Protein):", 
    paste(overlap_all_down, collapse = ", "), "\n")
```


