---
title: "Gene enrichment analysis"
author: "Simon Haugaard & Ali Altıntaş"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 3
    number_sections: true
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Gene enrichment analysis

## Required R libraries
```{r libraries}
# Install and load the necessary libraries using pacman
if (!require("pacman")) install.packages("pacman")  
pacman::p_load("magrittr",                          
               "data.table",                        
               "clusterProfiler",                   
               "jsonlite",                          
               "httr",                             
               "stringi", "ggplot2")    
```


## Read data
```{r read_data}
# Read differential gene expression (DGE) results from a specified file path
dge_file <- "../../01_dge/output/dge_results.tsv.gz"
dge <- fread(dge_file)  # Read in the DGE results file using fread

# Add a new column to categorize upregulated and downregulated genes based on log fold-change (logFC)
dge[, Direction := ifelse(logFC > 0, "up", "down")]

# Read gene annotation data from a specified file path
geneinfo_file <- "../../../data/gene_annotation/horse_gene_annotation.tsv.gz"
geneinfo <- fread(geneinfo_file)

# Set new column names for better clarity
setnames(geneinfo, new = c("ENSEMBL", "ENSEMBLv", "Description_detailed", 
                           "Chr", "Start", "End", "Strand", "GENENAME", "ENTREZID", "Description"))

# Remove unwanted columns and duplicated entries based on ENSEMBL IDs
geneinfo <- geneinfo[, c("ENSEMBLv", "Description_detailed") := NULL]  # Remove columns not needed
geneinfo <- geneinfo[!duplicated(ENSEMBL), ]                           # Retain only unique entries

# Extract gene names (GENENAME) from DGE data, assuming "GeneName" column exists in DGE file
dge[, GENENAME := sub(".*_", "", GeneName)]

# Merge DGE data with gene annotation data based on GENENAME
# The merge is performed to align DGE results with additional annotation information
dge <- merge(dge, geneinfo[, .(ENSEMBL, GENENAME, ENTREZID)], by = "GENENAME", all.x = TRUE)
dge <- dge[!is.na(ENSEMBL), ]

# Reorder columns to prioritize key identifiers (ENSEMBL, GENENAME, ENTREZID)
setcolorder(dge, c("ENSEMBL", "GENENAME", "ENTREZID", setdiff(names(dge), c("ENSEMBL", "GENENAME", "ENTREZID"))))

# Convert ENTREZID column to character for downstream compatibility
dge[, ENTREZID := as.character(ENTREZID)]

# Filter the DGE results for specific relevant contrasts of interest
dge <- dge[Contrast %in% c("AF_vs_Sham_RA", "AF_vs_Sham_LA", 
                           "Metformin_vs_AF_RA", "Metformin_vs_AF_LA", 
                           "AverageTreatmentEffect", "AverageDiseaseEffect")]

# Keep only entries with valid Entrez IDs for enrichment analysis
dge_entrez <- dge[!is.na(ENTREZID),]

# Define statistical cutoffs for differential gene expression and enrichment analysis
dge_cut <- 0.05        # Significance cutoff for differential gene expression (p-value)
enrich_cut <- 0.05     # Significance cutoff for enrichment analysis (q-value)


```

The analysis will be performed according to following statistical cutoffs.

- *Differential gene expression:* p-value < `r dge_cut`
- *Enrichment analysis:* adjusted p-value (Q-value) < `r enrich_cut`

# GO
```{r go}
# Read & Clean Gene Ontology (GO) Data
# The GO file contains the gene-to-GO term mappings necessary for enrichment analysis
go_file <- "../../../data/gene_annotation/horse_GO.tsv.gz"
go <- fread(go_file)

# Rename the columns for clarity and consistency
setnames(go, old = c("Gene stable ID", "GO term accession", "GO term name", "GO domain"),
         new = c("ENSEMBL", "GOID", "Description", "GOdomain"))

# Filter out rows with missing GO domain information
go <- go[GOdomain != "", ]

# Split GO mappings by domain (e.g., Biological Process, Molecular Function) into separate lists
term2gene <- split(go[,c("GOID", "ENSEMBL")], f = go$GOdomain)
term2name <- split(go[,c("GOID", "Description")], f = go$GOdomain)
```

## ORA 
```{r go_ora}
# Split differential gene expression (DGE) data into separate lists by contrast
dge_split <- split(dge, f = dge$Contrast)

# Initialize an empty list to store GO ORA results
go_ora <- list()

# Loop over gene sets ("all", "up", and "down"), GO domains, and contrasts to perform enrichment analysis
# Loop over gene sets ("all", "up", "down"), GO domains, and contrasts to perform enrichment analysis
for (k in c("all", "up", "down")) {
  for (j in names(term2gene)) {
    for (i in names(dge_split)) {
      message(paste0("Running GO ORA for '", i, "'", " within ", j, " ('", k, "' genes)"))
      
      # Select genes based on direction ("all", "up", "down") and significance cutoff
      if (k == "all") {
        gene <- dge_split[[i]][adj.P.Val < dge_cut, ENSEMBL]
      } else if (k == "up") {
        gene <- dge_split[[i]][adj.P.Val < dge_cut & Direction == "up", ENSEMBL]
      } else if (k == "down") {
        gene <- dge_split[[i]][adj.P.Val < dge_cut & Direction == "down", ENSEMBL]
      }
      
      # Skip if no differentially expressed genes (DEGs) found
      if (length(gene) == 0) {
        warning("No DEG found, skipping...\n")
        next
      }
      
      # Define the universe of genes to use as background in enrichment analysis
      universe <- dge_split[[i]][, ENSEMBL]
      
      # Perform GO enrichment analysis using the `enricher` function
      enrichment_result <- enricher(
        gene = gene,
        TERM2GENE = term2gene[[j]],
        TERM2NAME = term2name[[j]],
        pAdjustMethod = "BH",
        pvalueCutoff = 0.05,
        qvalueCutoff = 0.05,
        minGSSize = 10,
        maxGSSize = 500,
        universe = universe
      )
      
      # Check if the enrichment_result is NULL and skip if so
      if (is.null(enrichment_result)) {
        warning(paste0("No gene sets have size between 10 and 500 for '", i, "' in '", j, "' (", k, " genes). Skipping...\n"))
        next
      }
      
      # Store results and annotate them
      go_ora[[i]][[j]][[k]] <- enrichment_result@result
      go_ora[[i]][[j]][[k]]$Contrast <- i
      go_ora[[i]][[j]][[k]]$Database <- j
      go_ora[[i]][[j]][[k]]$Direction <- k
    }
  }
}


# Combine and convert the list of results into a single data table
go_ora_res <- do.call(rbind, go_ora %>% unlist(recursive=FALSE) %>% unlist(recursive = F)) %>% setDT

# Display the first few rows of the results for inspection
head(go_ora_res)

# Filter for significant terms based on the adjusted q-value cutoff
go_ora_res_significant <- go_ora_res[qvalue < enrich_cut, ]

# Save results to disk (commented out to avoid overwriting during testing)
fwrite(x = go_ora_res, file = "../output/GO_ora_sva.tsv.gz", sep = "\t")
go_ora_split <- split(go_ora_res, f = go_ora_res$Contrast)
openxlsx::write.xlsx(x = go_ora_split, file = "../output/GO_ora_sva.xlsx", asTable = TRUE)

# Remove the "all" category for plotting purposes
filtered_go_ora_res <- go_ora_res %>% filter(Direction != "all")

# Function to generate dot plots for enrichment results for specific contrasts
plot_dotplot_for_contrast <- function(data, contrasts, topn = 10, topn.pref = "qval", qcut = 0.05, nchar = 60, 
                                      direction = "Direction", group = "Contrast", dot = "GeneRatio", qval = "qvalue", 
                                      term.id = "ID", term.name = "Description") {
  for (contrast in contrasts) {
    # Filter data for the given contrast
    filtered_data <- data %>% filter(Contrast == contrast)
    
    # If data is available for the contrast, plot it
    if (nrow(filtered_data) > 0) {
      print(aamisc::dotplotEnrich(dt = filtered_data, 
                                  topn = topn, 
                                  topn.pref = topn.pref, 
                                  qcut = qcut, 
                                  nchar = nchar, 
                                  direction = direction, 
                                  group = group, 
                                  dot = dot,  # Ensure this matches the column name
                                  qval = qval, 
                                  term.id = term.id, 
                                  term.name = term.name))
    } else {
      # Display a message if no data is available for the contrast
      message(paste("No data for contrast:", contrast))
    }
  }
}

# Define a list of contrasts to plot
contrasts <- c("AF_vs_Sham_LA", "AF_vs_Sham_RA", "Metformin_vs_AF_LA",  "Metformin_vs_AF_RA","AverageDiseaseEffect")

# Generate dot plots for the defined contrasts
plot_dotplot_for_contrast(data = filtered_go_ora_res, contrasts = contrasts)

## Combined AF effects plot
# Function to generate a combined dot plot for specified contrasts with enhanced text
plot_combined_dotplot <- function(data, contrasts, topn = 10, topn.pref = "qval", qcut = 0.05, nchar = 60, 
                                  direction = "Direction", group = "Contrast", dot = "GeneRatio", qval = "qvalue", 
                                  term.id = "ID", term.name = "Description") {
  
  # Filter the data for the specified contrasts
  filtered_data <- data %>% filter(Contrast %in% contrasts)
  
  # Check if data is available for the contrasts
  if (nrow(filtered_data) > 0) {
    # Create the dot plot using aamisc::dotplotEnrich
    dot_plot <- aamisc::dotplotEnrich(dt = filtered_data, 
                                      topn = topn, 
                                      topn.pref = topn.pref, 
                                      qcut = qcut, 
                                      nchar = nchar, 
                                      direction = direction, 
                                      group = group, 
                                      dot = dot, 
                                      qval = qval, 
                                      term.id = term.id, 
                                      term.name = term.name) +
      # Enhance text elements for publication
      theme(
        axis.text.x = element_text(angle = 45, hjust = 1, size = 12, color = "black"),  # Rotate and adjust x-axis labels
        axis.text.y = element_text(size = 12, color = "black"),  # Increase y-axis label size
        axis.title = element_text(size = 14, face = "bold"),  # Bold and increase axis title size
        plot.title = element_text(hjust = 0.5, size = 16, face = "bold"),  # Center and style the plot title
        legend.title = element_text(size = 13, face = "bold"),  # Bold legend title
        legend.text = element_text(size = 11)  # Adjust legend text size
      ) +
      labs(
        title = "GO ORA Dot Plot for AF vs Sham (LA & RA)",  # Add a descriptive title
        x = "GO Terms",  # Label x-axis
        y = "-log10(q-value)"  # Label y-axis
      )
    
    # Print the dot plot to the R plotting window
    print(dot_plot)
  } else {
    # Display a message if no data is available for the contrasts
    message("No data available for the specified contrasts.")
  }
}

# Define the contrasts you want to combine
contrasts_to_plot <- c("AF_vs_Sham_LA", "AF_vs_Sham_RA")

# Generate and preview the dot plot in your R plotting window
plot_combined_dotplot(data = filtered_go_ora_res, contrasts = contrasts_to_plot)

# Save the plot using `ggsave` with matched dimensions
ggsave(filename = "../output/combined_dotplot.png", plot = last_plot(), 
       width = 8, height = 5, dpi = 300)

```


## GSEA
```{r go_gsea}
# Initialize go_gsea list and remove NAs
go_gsea <- list()


# Loop through each term2gene and dge_split combination
for (j in names(term2gene)){
    for (i in names(dge_split)){
        message(paste0("Running GO GSEA for '", i, "'", " within ", j))
        
        # Prepare geneList
        geneList <- dge_split[[i]]$logFC
        names(geneList) <- dge_split[[i]]$ENSEMBL
        geneList <- sort(geneList, decreasing = TRUE)
        
        # Run GSEA
        gsea_result <- GSEA(geneList = geneList, 
                            TERM2GENE = term2gene[[j]], 
                            TERM2NAME = term2name[[j]], 
                            pvalueCutoff = enrich_cut,
                            minGSSize = 10, 
                            maxGSSize = 500, 
                            eps = 0, 
                            nPermSimple = 10000)@result 
        
        # Check if the result is not empty
        if (nrow(gsea_result) > 0) {
            gsea_result$Contrast <- i
            gsea_result$Database <- j
        } else {
            gsea_result <- data.frame(Contrast = character(), Database = character())
        }
        
        # Store the result in the list
        go_gsea[[i]][[j]] <- gsea_result
    }
}

go_gsea_res <- do.call(rbind, go_gsea %>% unlist(recursive=FALSE)) %>% setDT
head(go_gsea_res)
go_gsea_res[qvalue < enrich_cut,]
go_gsea_res[, direction := ifelse(NES < 0, "down", "up")]

#Output data
go_gsea_res_split <- split(go_gsea_res, f = go_gsea_res$Contrast)

openxlsx::write.xlsx(x = go_gsea_res_split, file = "../output/GO_gsea_sva.xlsx", asTable = TRUE)
fwrite(x = go_gsea_res_split, file = "../output/go_gsea_sva.tsv.gz", sep = "\t")

# Function to plot dot plots for each contrast
plot_dotplot_for_contrast <- function(data_split, contrasts, topn = 10, topn.pref = "qval", qcut = enrich_cut, nchar = 60, 
                                      direction = "direction", group = "Contrast", dot = "NES", qval = "qvalue", 
                                      term.id = "ID", term.name = "Description") {
  for (contrast in contrasts) {
    if (contrast %in% names(data_split)) {
      filtered_data <- data_split[[contrast]]
      if (nrow(filtered_data) > 0) {
        print(aamisc::dotplotEnrich(dt = filtered_data, 
                                    topn = topn, 
                                    topn.pref = topn.pref, 
                                    qcut = qcut, 
                                    nchar = nchar, 
                                    direction = direction, 
                                    group = group, 
                                    dot = dot, 
                                    qval = qval, 
                                    term.id = term.id, 
                                    term.name = term.name))
      } else {
        message(paste("No data for contrast:", contrast))
      }
    } else {
      message(paste("Contrast not found:", contrast))
    }
  }
}

#Plotting
contrasts <- c("AF_vs_Sham_RA", "AF_vs_Sham_LA", "Metformin_vs_AF_RA", "Metformin_vs_AF_LA", "AverageDiseaseEffect", "AverageTreatmentEffect")
plot_dotplot_for_contrast(data_split = go_gsea_res_split, contrasts = contrasts)

```

### Output
```{r}
# Combine GSEA results into a single data frame
go_gsea_res_combined <- do.call(rbind, lapply(go_gsea, function(x) do.call(rbind, x)))

# Convert to a data frame if needed
go_gsea_res_combined <- as.data.frame(go_gsea_res_combined)
go_gsea_res_combined$Direction <- ifelse(go_gsea_res_combined$NES > 0, "up", "down")

# Function to generate a combined dot plot for specified contrasts with enhanced text
plot_combined_dotplot_gsea <- function(data, contrasts, topn = 10, topn.pref = "qval", qcut = 0.05, nchar = 60, 
                                       direction = "Direction", group = "Contrast", dot = "NES", qval = "qvalue", 
                                       term.id = "ID", term.name = "Description", title = "Combined GSEA Dot Plot") {
  
  # Filter the data for the specified contrasts
  filtered_data <- data %>% filter(Contrast %in% contrasts)
  
  # Check if data is available for the contrasts
  if (nrow(filtered_data) > 0) {
    # Create the dot plot using aamisc::dotplotEnrich
    dot_plot <- aamisc::dotplotEnrich(dt = filtered_data, 
                                      topn = topn, 
                                      topn.pref = topn.pref, 
                                      qcut = qcut, 
                                      nchar = nchar, 
                                      direction = direction, 
                                      group = group, 
                                      dot = dot, 
                                      qval = qval, 
                                      term.id = term.id, 
                                      term.name = term.name) +
      # Enhance text elements for publication
      theme(
        axis.text.x = element_text(angle = 45, hjust = 1, size = 12, color = "black"),  # Rotate and adjust x-axis labels
        axis.text.y = element_text(size = 12, color = "black"),  # Increase y-axis label size
        axis.title = element_text(size = 14, face = "bold"),  # Bold and increase axis title size
        plot.title = element_text(hjust = 0.5, size = 16, face = "bold"),  # Center and style the plot title
        legend.title = element_text(size = 13, face = "bold"),  # Bold legend title
        legend.text = element_text(size = 11)  # Adjust legend text size
      ) +
      labs(
        title = title,  # Add a descriptive title
        x = "GO Terms",  # Label x-axis
        y = "Normalized Enrichment Score (NES)"  # Label y-axis for GSEA
      )
    
    # Print the dot plot to the R plotting window
    print(dot_plot)
  } else {
    # Display a message if no data is available for the contrasts
    message("No data available for the specified contrasts.")
  }
}

# Define the contrasts you want to combine for GSEA
contrasts_to_plot <- c("Metformin_vs_AF_LA", "Metformin_vs_AF_RA")

# Generate and preview the dot plot in your R plotting window
plot_combined_dotplot_gsea(data = go_gsea_res_combined, 
                           contrasts = contrasts_to_plot, 
                           title = "Combined GSEA Dot Plot for Metformin Effects (LA & RA)")

# Save the plot using `ggsave` with matched dimensions
ggsave(filename = "../output/combined_gsea_dotplot_metformin.png", plot = last_plot(), 
       width = 10, height = 6, dpi = 300)
```


# KEGG
## ORA
```{r kegg_ora}
# This section performs KEGG pathway enrichment analysis using overrepresentation analysis (ORA).
KEGG_ora_upDown <- compareCluster(ENTREZID ~ Contrast + Direction, 
                             data          = dge_entrez[adj.P.Val < dge_cut,],
                             fun           = "enrichKEGG",
                             universe      = dge_entrez[, ENTREZID],
                             organism      = "ecb",
                             pAdjustMethod = "BH",
                             pvalueCutoff  = enrich_cut,
                             qvalueCutoff  = enrich_cut)@compareClusterResult %>% setDT

KEGG_ora <- rbindlist(list(KEGG_ora_upDown))
fwrite(x = KEGG_ora, file = "../output/kegg_ora.tsv.gz", sep = "\t")

KEGG_ora_split <- split(KEGG_ora, f = KEGG_ora$Contrast)
openxlsx::write.xlsx(x = KEGG_ora_split, file = "../output/kegg_ora_sva.xlsx", asTable = TRUE)

# Plot results
aamisc::dotplotEnrich(dt = KEGG_ora, 
                      topn = 10, 
                      topn.pref = "qval", 
                      qcut = enrich_cut, 
                      nchar = 60, 
                      direction = "Direction", 
                      group = "Contrast", 
                      dot = "GeneRatio", 
                      qval = "qvalue", 
                      term.id = "ID",
                      term.name = "Description")

# Function to plot dot plots for each contrast for KEGG enrichment results
plot_dotplot_for_contrast_kegg <- function(data, contrasts, topn = 10, topn.pref = "qval", qcut = enrich_cut, nchar = 60, 
                                           direction = "Direction", group = "Contrast", dot = "GeneRatio", qval = "qvalue", 
                                           term.id = "ID", term.name = "Description") {
  for (contrast in contrasts) {
    filtered_data <- data %>% filter(Contrast == contrast)
    if (nrow(filtered_data) > 0) {
      print(aamisc::dotplotEnrich(dt = filtered_data, 
                                  topn = topn, 
                                  topn.pref = topn.pref, 
                                  qcut = qcut, 
                                  nchar = nchar, 
                                  direction = direction, 
                                  group = group, 
                                  dot = dot, 
                                  qval = qval, 
                                  term.id = term.id, 
                                  term.name = term.name))
    } else {
      message(paste("No data for contrast:", contrast))
    }
  }
}

# Example usage
# Plot
filtered_KEGG_ora <- KEGG_ora %>% 
  filter(Direction != "all")

plot_dotplot_for_contrast_kegg(data = filtered_KEGG_ora, contrasts = "AF_vs_Sham_LA")
```


## GSEA 
```{r kegg_gsea}
# This section performs Gene Set Enrichment Analysis (GSEA) on KEGG pathways using the `gseKEGG` function.
# The analysis is performed for each contrast separately using ENTREZ IDs.
dge_split <- split(dge_entrez, f = dge_entrez$Contrast)
kegg_gsea <- list()

for (i in names(dge_split)){
  geneList <- dge_split[[i]]$logFC
  names(geneList) <- dge_split[[i]]$ENTREZID
  geneList <- sort(geneList, decreasing = TRUE)
  
  kegg_gsea[[i]] <-  gseKEGG(geneList     = geneList,
                        organism     = 'ecb',
                        pvalueCutoff = enrich_cut,
                        minGSSize = 10, 
                        maxGSSize = 500, 
                        eps = 0, 
                        nPermSimple = 10000, # TODO: set it to 1 million for robust results
                        verbose      = TRUE
                        )@result %>% setDT
  kegg_gsea[[i]][, Direction := ifelse(test = NES > 0, yes = "up", no = "down"), ]
  kegg_gsea[[i]]$Contrast <- i
}

aamisc::dotplotEnrich(dt = kegg_gsea[["AF_vs_Sham_LA"]], 
                      topn = 10, 
                      topn.pref = "qval", 
                      qcut = enrich_cut, 
                      nchar = 60, 
                      direction = "Direction", 
                      group = "Contrast", 
                      qval = "qvalue", 
                      dot = "NES", 
                      term.id = "ID",
                      term.name = "Description")
# Function to plot dot plots for each contrast for KEGG GSEA results
plot_dotplot_for_contrast_kegg_gsea <- function(data_split, contrasts, topn = 10, topn.pref = "qval", qcut = enrich_cut, nchar = 60, 
                                                direction = "Direction", group = "Contrast", dot = "NES", qval = "qvalue", 
                                                term.id = "ID", term.name = "Description") {
  for (contrast in contrasts) {
    if (contrast %in% names(data_split)) {
      filtered_data <- data_split[[contrast]]
      if (nrow(filtered_data) > 0) {
        print(aamisc::dotplotEnrich(dt = filtered_data, 
                                    topn = topn, 
                                    topn.pref = topn.pref, 
                                    qcut = qcut, 
                                    nchar = nchar, 
                                    direction = direction, 
                                    group = group, 
                                    qval = qval, 
                                    dot = dot, 
                                    term.id = term.id, 
                                    term.name = term.name))
      } else {
        message(paste("No data for contrast:", contrast))
      }
    } else {
      message(paste("Contrast not found:", contrast))
    }
  }
}

# Plot
plot_dotplot_for_contrast_kegg_gsea(data_split = kegg_gsea, contrasts = contrasts)
```

# Session Info
```{r session_info}
sessionInfo()
```
